### PR TITLE
avoid duplication of source strings

### PR DESCRIPTION
**Pull request by the8472**
Currently script sources get cloned into every sandbox in which a script runs via the `GM_info` structure. Since script sources can be fairly large this results in quite some memory overhead.

Additionally in e10s the scripts are passed down via IPC as new string instances every time they're needed, thus resulting in duplication even before the `GM_info` instances are created.

I've solved that by de-duping identical string instances in `IPCScript` and by installing a cross-compartment getter for the `GM_info.scriptSource` property.

Should lead to significant memory reductions, especially for large scripts and/or scripts that get loaded into every tab and frame.

### CODE REVIEW

File: modules/ipcscript.js [hunk 1]
```diff
@@ -48,6 +50,21 @@ function IPCScript_getMetaStr() {
...
+  // alternative approaches:
+  // a) beam down IPC script instances only once via the parent process manager every time they change.
+  // b) creating resource URIs and loading user scripts straight from URIs might also be more elegant
+  //    see: https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Using#Programmatically_adding_aliases
+  var old = instances.get(this.uuid);
+
+  if(old == this.textContent)
+    this.textContent = old
+  else
+    instances.set(this.uuid, this.textContent)
```
Comment (arantius) [inline modules/ipcscript.js]: When are instances unset?

**Comment by arantius** [comment]
Please don't skip optional braces or semicolons.

File: modules/ipcscript.js (hunk 1 above)
Comment (the8472) [inline modules/ipcscript.js]: on the assumption that uninstalling scripts is a rather rare use-case and updating/modifying them is more likely it simply overwrites them based on UUID. removal would require additional communication with the parent process since the child process otherwise has no way of knowing.

Comment (arantius) [inline modules/ipcscript.js]: So before this change, the script string is in memory only while the script is executing/is in scope.  It is garbage collected once it goes out of scope (i.e. navigation to a different page, close the tab, or as soon as it finishes executing if it doesn't expose references to itself to content).

After this change, every script that has ever executed has a copy of its source stored permanently in memory.  When there are several children processes (there is only one today AIUI) there will be several copies, and none of them will ever go away until the browser is completely closed.

It's not clear to me whether this new approach is better overall in terms of memory usage.

Comment (the8472) [inline modules/ipcscript.js]: Mhh, good point, I was basing things on behavior I observed in my profile, where everything is fairly long-lived anyway.

The scenario you outline seems plausible for other types of script. I'll see If i can implement option a)

But for that one bit of information would be useful: Can I rely on all changes to `Script` instances notifying `service.config.addObserver()`

Comment (arantius) [inline modules/ipcscript.js]: Changes to scripts should always be notifying the Config.

But option A sounds a lot like "store the content of every script in memory forever" again, in a slightly different arrangement.  I'm curious what we can do with lazy getters and either sync messages or however that resource URI thing would work.

Comment (the8472) [inline modules/ipcscript.js]: Hrrm, wait... you want scripts to be GCed once they're not used anymore by a child process.

I guess some weakmap magic + keeping instances alive as long as they're referenced by the respective sandboxes might work.

**Comment by the8472** [comment]
pushed new approach using lazy getters and loading via file:// uris. I didn't actually have to do any resource:// registration, mozilla seems to be doing about the right thing, sharing instances from the same path.

Scripts touching `textContent` will lead to duplicate strings for the lifetime of the sandboxes, but as long as they don't it saves memory.

Dependency loading also seems to work.

What still needs testing is resource-loading in e10s, none of my scripts use that.

**Comment by the8472** [comment]
resource loading should work now

**Comment by arantius** [comment]
Started looking into this, I'm not happy with it.  The whole concept of `fileXhr` seems brittle to me.  Mozilla has explained that they plan to continue e10s by making it impossible for the child process to (e.g.) access local files at all.  The fact that XHR happens to be a workaround to blockage of `nsIFile` doesn't stop this.  One day this change will (probably?) fail to work correctly, and we'll be left scrambling.

**Comment by the8472** [comment]
[This page](https://developer.mozilla.org/en-US/Firefox/Multiprocess_Firefox/Which_URIs_load_where) states that `file://` will always load in content, so I would expect this to work in the future too.

I think mozilla wants to prevent direct system calls to file handles (`open()` and the like), using XHR just passes the input stream to the content process AIUI, not the file handle itself.

So it doesn't seem obvious to me that they'll remove `file://` XHR

**Comment by the8472** [comment]
further reading: https://bugzilla.mozilla.org/show_bug.cgi?id=922481

**Comment by the8472** [comment]
also asked one of the devs:

```
<The_8472> billm, file: URIs will continue to work though, yes?
<billm> The_8472: sure
```

**Comment by arantius** [comment]
Is this implicit in #2259 and safe to ignore if that's merged?

**Comment by the8472** [comment]
Yes
